Each toolchain has custom set of compiler wrappers

karkhaz opened the pull request:
The compiler wrapper that sits in place of native tools now prints out
an error message and proceeds to invoke the corresponding toolchain tool.

Toolchains now declare what native tools should be overwritten with the
compiler wrapper using a YAML file. Each custom compiler wrapper, as
well as the custom Dockerfile, is generated from this declaration when
running ./tuscan.py build.

Commit:
Hybrid package creation

This commit enables the creation of hybrid packages---packages
containing both vanilla and toolchain-built data.

New stages
----------

The create_toolchain_repo stage is added. This creates an empty
package repository in a data container. This package repository is used
by the make_package stage:
  * when a package is successfully built, it is added to the package
    repository;
  * dependencies of a build are installed from the package repository
    before the build begins.

The install_bootstrap stage is added. This installs all vanilla packages
that are necessary to build all other packages (the bootstrap packages).
This container is committed after it finishes running, so that
containers running the make_package stage can use it as a base image.
This stage is run just once, since installing all the bootstrap packages
is a time-consuming process.

Local Arch Mirror
-----------------

Packages are now synced from a local arch mirror, rather than upstream.
This saves bandwidth and is more polite, as well as making the
experiment reproducible.

Hybrid package creation
-----------------------

The make_package and create_toolchain_repo stages now create hybrid
packages. When a package is built, the hybrid packages that it depends
on are installed, and the compiler uses toolchain-built libraries and
headers.

Logging
-------

The make_package, install_bootstrap and create_toolchain_repo stages
emits JSON-structured logs that can be parsed into HTML reports.

Bugfix: file exists in filesystem

Before this commit, the toolchain portion of hybrid packages would be
installed in a global location and not uninstalled. However, the package
database is local to each container. This means that containers
attempting to install a dependency whose files already existed on the
file system would fail.

This commit makes the installation location of toolchain files local to
each container, while keeping the toolchain archive global.

Compiler wrapper
----------------

This commit introduces a rudimentary compiler wrapper that replaces the
native GCC, ld and ar programs. It prints out the name of the program it
is replacing and then returns non-zero.

This is so that we can detect when toolchain builds are ignoring our
setting of CC and hardcoding the name of the native compiler.

Commit:
Instructions for adding toolchain & getting data

Added instructions for adding new toolchains, and getting sources and
binaries.

Also refactored setup.py to make it easier to add new toolchains.
Closes #45.

karkhaz committed:
make_package logs info on provided packages

The make_package stage now writes a list of packages that are provided
by its build to the build log. This is done so that we know what
packages were supposed to have been built by this build.

Any future build that depends on one of the packages will know which
build to blame if it cannot install some of its dependencies. This will
allow us to show a graph of what builds are causing lots of future
builds to fail.

This commit has necessitated refactoring out some common code to
utilities.py.

karkhaz committed:
Data post-processing skeleton

This commit allows the user to run `tuscan.py post` in order to
post-process JSON data that was dumped by the make_package stage.
Currently, no actual post-processing is done; the data is written as-is
to the `post` directory.

karkhaz committed:
Added schema for stage deps.yaml file

This commit adds a schema for the deps.yaml file included in each
'stages' directory. This schema serves for documentation of the file
format as well as validation, due to the fact that most of the keys are
optional.

karkhaz committed:
Added schema for data_containers.yaml

This commit adds a schema for the data_containers.yaml file, to serve as
validation and documentation of the file format.

karkhaz committed:
Error classification of log files

This commit introduces the ability to classify each line of the build
output as indicating some error. The post-processing pass now decorates
each line of the build output with its error category.

* classification_patterns.yaml contains the patterns that we attempt to
  match each line of the build output against, as well as what kind of
  error that line indicates if it matches.

* schemata.py has been updated to reflect the extra fields in the
  post-processed JSON.

karkhaz committed:
HTML report generation for individual builds

This commit introduces the `tuscan.py html` command, which generates a
HTML report for each post-processed build log.

karkhaz committed:
Flagging return codes of configure script logs

Post-processed data now has the following two changes:

* The overall data has a required key called "config_success", which
  indicates whether the configure invocations for this build were all
  successful. This is so that we can distinguish between builds that
  failed during and after the configure stage.

* Each log fragment has an optional key called "config_success", which
  exists if that log fragment is a configure log (dumped by autoconf or
  the like). This is so that we can ignore 'errors' that happened during
  configure invocations that returned successfully, i.e. where the
  'errors' were actually expected.

karkhaz committed:
All output files rooted at single output directory

The following three directories, which were created at the top level,
will now be created in a single top-level directory called output:

- results:  the JSON results of running the experiment
- post:     the post-processed JSON results
- html:     HTML reports generated from post-processed results

This is so that these directories can be stored on a separate high-
performance filesystem with write barriers and journaling disabled,
mounted on the output directory, for increased system responsiveness and
reduced disk wear.

karkhaz committed:
Missing source is detected earlier in build

This commit adds a check for source directories that don't exist on the
permanent source volume at all (as opposed to source that can't be
obtained for some other reason). It is an earlier check for missing
source than the checks we already had, and fails the build with a log
message rather than causing a crash.

karkhaz committed:
Dependencies of each build are logged for postproc

Each build now logs the packages that it depends on before beginning.
This will allow us to build a graph of what failing packages are
blocking other packages from building, and determining what packages are
the biggest blockers etc.

karkhaz committed:
HTML summary pages with sidebar

tuscan_html now produces summary pages that group lists of builds
satisfying certain criteria. These criteria are grouped hierarchically;
a sidebar on each summary page shows the tree structure of the
categories.

karkhaz committed:
Calculation of blockers/blocked-by in postprocess

This commit adds two new fields to the post-processed data structure:

- blocked_by: if this build was not attempted because some of its
  dependencies failed to build (FTB), this field will list those
  dependencies.

- blocks: if this build FTB even though all of its dependencies built
  successfully, then this build is said to be a "blocker". This field
  will list all builds that transitively depend on this one, and which
  would not have been built.

This commit will enable the HTML report to contain hyperlinks from
blocked builds to their blockers. It may be worth manually patching
blockers that block a large number of builds in order to get better
results.

karkhaz committed:
SLOC information dumped during build

The make_package stage now dumps lines-of-code information into the
result data structure for a build. The result data structure now
includes a dictionary mapping languages to lines of code written in that
language for that build.

karkhaz committed:
Initial tuscan-figures code

There is now a ./tuscan.py figures command that generates on
post-processed data. No figures are produced at present.

karkhaz committed:
SLOC distribution figure

The `tuscan figures` command now generates a plot showing the
distribution of sizes of programs that have been built. For each order
of magnitude of lines of code (10 LOC, 100 LOC, ...) the graph shows how
many programs are at most that large.

karkhaz committed:
Addition of musl toolchain

A clang+llvm+libcxx+musl toolchain is now supported, in addition to the
upstream and android toolchains.

The following changes have been made to the build infrastructure in
order to support this change:

- Toolchain-specific files for the install_bootstrap stage and
  make_package stage are now located in the directory
  $TOOLCHAIN/{install_bootstrap,make_package}, rather than being located
  directly under $TOOLCHAIN. This is because musl requires a slightly
  different Dockerfile for its install_bootstrap stage than the other
  toolchains, so each toolchain can now specify its own Dockerfile for
  the install_bootstrap stage.

- The `copy_files` field in deps.yaml files now supports globbing, in
  order to copy all files in a directory without knowing what they are
  called. This is because some of the toolchains have a Python setup
  script, while the musl toolchain uses a shell script.

karkhaz committed:
Allow user to specify toolchain to post-process

The user can now optionally specify a list of toolchains to
post-process, instead of post-processing the results of all toolchains
(which remains the default). This is because post-processing it taking a
large amount of main memory, and running the post-processing one
toolchain at a time can help the machine to not thrash.

karkhaz committed:
HTML sidebar filters out non-C(++) builds

The HTML report now excludes builds that do not contain any C or C++
code at all but the top-level.

karkhaz committed:
New HTML summary for unclassified errors

A HTML summary is now generated for builds that had no errors that we
were able to classify.

karkhaz committed:
Vanilla-failed builds ignored on HTML summary

Statistics about alternative toolchains no longer take into account any
build that failed on vanilla. There is still a top-level summary that
shows builds that passed and failed on vanilla.

karkhaz committed:
More accurate build count on HTML summary pages

Summary pages can display the total number of builds on that page.
Previous to this commit, summary pages that displayed builds for more
than one toolchain would count each build once for each toolchain, i.e.
tripling the apparent number of builds if there are three toolchains on
display.

This commit changes that behaviour so that the "total" number of builds
is the number of builds divided by the number of toolchains on display.

karkhaz committed:
Improved memory efficiency of blocker post-process

Post-processed results are no longer stored in an inter-process data
structure while waiting to have their blockers calculated, as this was a
huge waste of memory. Instead, results are written to disk after all
post-processing except for blocker analysis is done. When all results
have been written, only the necessary data for blocker analysis is re-
loaded from disk; the on-disk results are then updated with the blocker
results.

Due to this improvement, the blocker analysis has been re-enabled by
default.

karkhaz committed:
HTML display of blockers

Blockers for each toolchain are displayed in the summary sidebar.
Packages which block others from building, and packages that have been
blocked from building, now contain a message on their individual pages.

karkhaz committed:
Bear is called on each build

The bear tool, which has been added as a submodule of this project, is
used to wrap each invocation of makepkg. The output is stored as part of
the JSON result structure.

petrhosek commented on tuscan/compiler_wrapper.c:
```diff
@@ -0,0 +1,61 @@
+    exit(1);
+  }
+
+  char toolchain_tool[sizeof(TOOLCHAIN_TOOL)] = TOOLCHAIN_TOOL;
```
No need to specify the length of the array. The array should be also `const`. Furthermore, since you're generating this program anyway, I'd avoid the macro too and just use:

``` c
const char *toolchain_tool = "{{ toolchain_bin }}/{{ toolchain_program }}"
```

petrhosek commented on tuscan/compiler_wrapper.c:
```diff
@@ -0,0 +1,61 @@
+  char *new_argv[argc];
+  unsigned i;
+  new_argv[0] = toolchain_tool;
+  for(i = 1; i < argc; i++){
```
Since you're using C99, you can inline the declaration of the iterator:

``` c
for (int i = 1; i < argc; i++)
```

petrhosek commented on tuscan/compiler_wrapper.c:
```diff
@@ -0,0 +1,61 @@
+#define RANDOM_PATH_LENGTH 20
+#define TOOLCHAIN_TOOL "{{ toolchain_bin }}/{{ toolchain_program }}"
+
+int main(unsigned argc, char *argv[]){
```
`argc` should be `int` (I know, it's counterintuitive but that's what C standard says), also in C programs `argv` is usually `char **`; `char* argv[]` is more common in C++ (even though they're equivalent).

petrhosek commented on tuscan/compiler_wrapper.c:
```diff
@@ -0,0 +1,61 @@
+#define TOOLCHAIN_TOOL "{{ toolchain_bin }}/{{ toolchain_program }}"
+
+int main(unsigned argc, char *argv[]){
+  unsigned buf[RANDOM_PATH_LENGTH];
```
This is wrong, if you want buffer of `RANDOM_PATH_LENGTH` bytes, you need to use `char buf[RANDOM_PATH_LENGTH]`. In general, I'd avoid using `unsigned` as a type, if anything it should be `unsigned int`.

petrhosek commented on tuscan/compiler_wrapper.c:
```diff
@@ -0,0 +1,61 @@
+
+int main(unsigned argc, char *argv[]){
+  unsigned buf[RANDOM_PATH_LENGTH];
+  if (syscall(SYS_getrandom, buf, RANDOM_PATH_LENGTH, 0) == -1){
```
You can use `sizeof(buf)` here and avoid the need for the `RANDOM_PATH_LENGTH` macro.

petrhosek commented on tuscan/compiler_wrapper.c:
```diff
@@ -0,0 +1,61 @@
+    exit(1);
+  }
+  char f_name[32];
+  sprintf(f_name, "/tmp/tuscan-native-%d", abs((int)(*buf)));
```
First of all, you should be using `snprintf`, always. Second, that cast of the buffer content to int is super hacky. The cleaner way of doing is would be the following:

``` c
unsigned long long buf;
if (syscall(SYS_getrandom, &buf, sizeof(buf), 0) == -1) { ... }
char filename[64];
snprintf(filename, sizeof(filename), "/tmp/tuscan-native-%llu", buf);
```

However, since you're using to generate a temporary filename, I'd just use the `mkstemp` function from standard C library which already implements that functionality, so all of the above can be replaced with:

``` c
int fd = mkstemp("/tmp/tuscan-native-XXXXXX");
if (fd == -1) {
  perror("tuscan: mkstemp");
  return 1;
}
dprintf(fd, "{{ native_program }}\n");
if (close(fd) == -1) {
  perror("tuscan: close")
  return 1;
}
```

Please note that `mkstemp` already opens the file for you to avoid the race condition, but it returns a file descriptor, so you need to use `dprintf` and `close` instead of `fprintf` and `fclose`.

karkhaz commented on tuscan/compiler_wrapper.c:
```diff
@@ -0,0 +1,48 @@
+    return 1;
+  }
+
+  char *toolchain_tool = "{{ toolchain_bin }}/{{ toolchain_program }}";
```
I implemented all the changes you mentioned except for making this a `const char*`. This is because `new_argv` is required by `execv` to be of type `char * const[]` and not `const char * const[]`, and so assigning a `const char*` string to the array breaks the constness of that particular string.

`toolchain_tool` is also used as the _first_ parameter to `execv`, which does expect a `const char *`, but that's okay because it just gets upgraded to `const` when the function is invoked.

I could alternatively do

```
char *toolchain_tool = "{{ toolchain_bin }}/{{ toolchain_program }}";
const char *c_toolchain_tool = "{{ toolchain_bin }}/{{ toolchain_program }}";
char *new_argv[argc];
new_argv[0] = toolchain_tool;
for(int i = 1; i < argc; i++){
   new_argv[i] = argv[i];
  }
  execv(c_toolchain_tool, new_argv);
}
```

but I don't think that's necessary...

petrhosek replied:
This looks fine.

karkhaz committed:
Each toolchain has custom set of compiler wrappers

The compiler wrapper that sits in place of native tools now prints out
an error message and proceeds to invoke the corresponding toolchain tool.

Toolchains now declare what native tools should be overwritten with the
compiler wrapper using a YAML file. Each custom compiler wrapper, as
well as the custom Dockerfile, is generated from this declaration when
running ./tuscan.py build.

karkhaz commented:
(please ignore the fact that it's two commits, that's just because I just rebased this onto master after merging in another PR)

petrhosek commented on tuscan/compiler_wrapper.c:
```diff
@@ -0,0 +1,48 @@
+#include <stdio.h>
+#include <unistd.h>
+
+int main(int argc, char **argv){
```
Nit: there should be a space between `)` and `{`.

petrhosek commented on tuscan/wrapper_gen.py:
```diff
@@ -0,0 +1,103 @@
+            logging.error("File not found: '%s'" % f)
+            exit(1)
+
+    with open(os.path.join(make_package_dir, "binutil_transforms.yaml")) as f:
```
I'd consider using a different name for the file, `binutil` is confusing since it can refer to GNU Binutils (that's what I originally thought).

petrhosek commented on tuscan/wrapper_gen.py:
```diff
@@ -0,0 +1,103 @@
+        make_package_dir]))
+
+    dockerfile_temp = jinja.get_template("Dockerfile")
+    wrapper_temp = jinja.get_template("compiler_wrapper.c")
```
I'd probably omit the `compiler_` part from the name since the wrapper is used for other tools as well.

karkhaz committed:
No longer importing members from modules

All instances of "from ... import ..." have been replaced with a module
import and members are referenced by their fully-qualified names.

karkhaz committed:
Added additional packages to the provides array
